lib/Backend/LowerMDShared.cpp

```diff
@@ -5812,6 +5815,114 @@ LowererMD::GenerateCtz(IR::Instr * instr)
+    return xmmReg;
+}
+
+IR::RegOpnd* LowererMD::MaterializeConstFromBits(int bits, IRType type, IR::Instr* instr)
+{
+    IR::Opnd * regBits = IR::RegOpnd::New(TyInt32, m_func);
+    this->m_lowerer->InsertMove(regBits, IR::IntConstOpnd::New(bits, TyInt32, m_func), instr);
+    IR::RegOpnd * regConst = IR::RegOpnd::New(type, m_func);
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::MOVD, regConst, regBits, m_func));
+    return regConst;
+}
+
+IR::Opnd* LowererMD::GenerateTruncChecks(IR::Instr* instr)
+{
+    IR::LabelInstr * conversion = IR::LabelInstr::New(Js::OpCode::Label, m_func);
+    IR::LabelInstr * throwLabel = IR::LabelInstr::New(Js::OpCode::Label, m_func);
+    IR::Opnd* src1 = instr->GetSrc1();
+
+    IR::Opnd * src64 = nullptr;
+    if (src1->IsFloat32())
+    {
+        src64 = IR::RegOpnd::New(TyFloat64, m_func);
+        EmitFloat32ToFloat64(src64, src1, instr);
+    }
+    else
+    {
+        src64 = src1;
+    }
+
+    IR::RegOpnd* limitReg = nullptr;
+    if (instr->GetDst()->IsUInt32())
+    {
+        limitReg = IR::RegOpnd::New(TyFloat64, m_func);
+        IR::Instr* xor = IR::Instr::New(Js::OpCode::XORPS, limitReg, src64, src64, m_func);
+        instr->InsertBefore(xor); //min = 0
+        Legalize(xor);
+    }
+    else
+    {
+        limitReg = MaterializeDoubleConstFromInt(m_func->GetThreadContextInfo()->GetDoubleIntMinAddr(), instr);
+    }
+    m_lowerer->InsertCompareBranch(src64, limitReg, Js::OpCode::BrLt_A, throwLabel, instr);
+
+    if (instr->GetDst()->IsUInt32())
+    {
+        limitReg = IR::RegOpnd::New(TyFloat64, m_func);
+        IR::Opnd * twoTo31Float = MaterializeConstFromBits(TWO_31_FLOAT, TyFloat32, instr);
+        EmitFloat32ToFloat64(limitReg, twoTo31Float, instr);
+        IR::RegOpnd* maxIntDoubleReg = MaterializeDoubleConstFromInt(m_func->GetThreadContextInfo()->GetDoubleIntMaxAddr(), instr);
+        instr->InsertBefore(IR::Instr::New(Js::OpCode::ADDPD, limitReg, limitReg, maxIntDoubleReg, m_func));
+    }
+    else
+    {
+        limitReg = MaterializeDoubleConstFromInt(m_func->GetThreadContextInfo()->GetDoubleIntMaxAddr(), instr);
+    }
+
+    m_lowerer->InsertCompareBranch(limitReg, src64, Js::OpCode::BrGe_A, conversion, instr, true /*no NaN check*/);
+    instr->InsertBefore(throwLabel);
+    IR::Instr *throwInstr = IR::Instr::New(
+        Js::OpCode::RuntimeTypeError,
+        IR::RegOpnd::New(TyMachReg, m_func),
+        IR::IntConstOpnd::New(SCODE_CODE(VBSERR_Overflow), TyInt32, m_func),
+        m_func);
+    instr->InsertBefore(throwInstr);
+    this->m_lowerer->LowerUnaryHelperMem(throwInstr, IR::HelperOp_RuntimeTypeError);
+    //no jump here we aren't coming back
+
+    instr->InsertBefore(conversion);
+    return src64;
+}
+
+void
+LowererMD::GenerateTruncWithCheck(IR::Instr * instr)
+{
+    Assert(AutoSystemInfo::Data.SSE2Available());
+
+    IR::Opnd* src64 = GenerateTruncChecks(instr); //converts src to double and checks if  MIN <= src <= MAX
```

Comment:
@MikeHolman 

Agreed. For the unsigned case, we would still have **3** checks in the worst case scenario and **2** checks in the best case?  
 
We check for an overflow `80000000H`  (1st check) 
* No overflow
  * Check if the source is positive (2 checks)
* Overflow 
  * Check if the source is positive, 2 checks (possibly **MIN_INT**-1)
  * Check if the source is less than **MAX_UINT**, 3 checks

Does this sound right?